### PR TITLE
Handle 'is' operators and patterns in definite assignment

### PR DESCRIPTION
**Pull request by RikkiGibson**
Related to #51463

Implements the following section of the spec: https://github.com/dotnet/csharplang/blob/main/proposals/improved-definite-assignment.md#is-operator-and-is-pattern-expressions

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs
Comment (alrz) [inline src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs:4323]: Should any of these work with `e is not null is true` and similar?

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:1045]: Oops, this makes me realize that I meant to only evaluate the right side if the left side returned non-`null`.

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs:4323]: Yes :)

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:967]: Small nit: this could still be at the start of the function.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:970]: >GetTopLevelNullTest [](start = 24, length = 19)

It looks like this is only used here. Consider making it a local function.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:1055]: This should probably throw on unexpected patterns, not return.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:1022]: >Get [](start = 22, length = 3)

Consider `IsTopLevelNullTest`, rather than `Get`.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:982]: >intentionally [](start = 70, length = 13)

Consider elaborating. This is mostly an opaque comment that doesn't explain why we're doing this.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:983]: >ConstantValue: { IsBoolean: true, BooleanValue: var value } [](start = 59, length = 59)

Why only bool patterns?

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:982]: >negated [](start = 137, length = 7)

This feels wrong. `is not true` should have the same behavior as `is false`.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:60]: I don't understand this change?

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs:4120]: Yeah, I really don't like this. `not` needs to be symmetrical.

**Comment by 333fred** [comment]
Done review pass (commit 5). Didn't look through all the tests, just a few.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:60]: This is something I should have changed at the same time that I made bool constants put the walker into a conditional state. Without this change, we crash in test `FlowTests.IsBoolConstant_03` because we enter a split state after visiting one of the binary pattern operands, and then attempt to visit the other binary pattern operand while in a split state.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: Done

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: Done

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: Done. Have manually enumerated the pattern kinds, which may help us catch it and decide what to do when a new pattern kind is added to the language.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: Done

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: Have included an example in the comment to try and make it more clear.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: I realized this was what the spec requires as written, and didn't feel interested in the scenarios that were missed. Opened dotnet/csharplang#4664 and changed the implementation here.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: Fixed

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs]: Fixed

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs:4541]: >int [](start = 4, length = 3)

Consider a version where this returns `int?`.

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs:4541]: And specifically include a case where you test against `null`.

---
In reply to: [615140930](https://github.com/dotnet/roslyn/pull/52616#discussion_r615140930) [](ancestors = 615140930)

**Comment by 333fred** [comment]
Done review pass (commit 7).

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs]: Have expanded the test to cover these scenarios.

**Review by 333fred — APPROVED** [review]
LGTM (commit 8)

**Comment by RikkiGibson** [comment]
@jcouv Please take a look when you have the chance, need a second review.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:1098]: nit: should this be pattern.Kind? (same above)

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:1023]: the local function probably needs a negation (either changing its name, or changing return values).

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:1027]: I'm not sure when `expr is T` results in a `BoundIsOperator` as opposed to the new pattern representation (`BoundIsPatternExpression`). We may need to handle that case too.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:970]: I assume that we can assert that the state is not conditional. Consider adding assertion

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:1027]: Ah, I see that you handled that already (`VisitIsOperator` below). Never mind :-)

---
In reply to: [616348015](https://github.com/dotnet/roslyn/pull/52616#discussion_r616348015) [](ancestors = 616348015)

**Review by jcouv — APPROVED** [review]
Done with review pass (iteration 8). Only minor comment

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 8). Only minor comment

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 8). Only minor comment

**Comment by jcouv** [comment]
Something is messed up with GitHub. I left a comment (ie. didn't approve yet) and somehow the PR shows an approval...
I'll clear my review to avoid confusion.

image
image

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: Done

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: Done

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:970]: Done

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 9)

**Pull request merged**
